Spark 130 -- Cofiguration overhaul.

theunkn0wn1 opened the pull request:
This pull request completley redoes how Mecha handles configurable settings.

# High level changes
This re-implementation uses [`pluggy`](https://github.com/pytest-dev/pluggy) as its core infrastructure.


1. This package is already in our development dependencies ( `Pytest` depends on it amongst other modules ), meaning its already present in our development environments.
2. `Pluggy` prevents us from re-inventing the wheel several times over, il get tot his more later.
3. `Pluggy` lays the foundation for **plugins**
4. `Pluggy` is extensible

Pluggy is designed to enable plugins, making this approach possibly the most extensible solution to the problem of event propagation.

## Existing dependency
- By using an existing dependency, We know it doesn't break anything when introduced.
- It is a well tested, well developed package made by the same people as our `pytest` test runner.

## Re-inventing the wheel and alternative implementations
I have considered other implementations for the configuration overhaul including but not limited to abstract / Base classes and decorators.


## objectives
 - [x] the new system must be extensible
    - Seems simple enough to extend with other event types
- [x] the new system should support plugins 
   - pluggy is *designed* for plugins
- [x] the new system must support reloading of the configuration file, aka `rehashing`
    - Implemented with literal file hashing to boot
- [x] the new system must support validating the configuration file
   - done through plugin events.

### subclasses
The issue with the usage of subclasses, a la Offline Awareness, is that  i did not find this approach to be sufficiently extensible. 

OA's implementation depends on a Base class that keeps shallow references to its children.
It does this by requiring children to invoke the base class's constructor via a super call.
```py
class Aware(OfflineAwareABC):
    async def on_online(self):
        ...

    async def on_offline(self):
         ...
    def __init__(self):
        ...
        super().__init__()
```


On its own, this works fine. However, once we consider the underlying interface, the problems become obvious in the context of extensibility.

1. it uses class-bound attributes
- By using class-bound attributes, out of necessity any subclass of this object will have these attributes too.
-  specialized subclasses would need their own unique storage object
2. machinery can't be reused between specialized event ABCs, **leading to duplication
 - each specialized event ABC would need to essentially copy/paste the machinery, due to the cross-ABC state.
 - reinvention of wheel.
3. cross-ABC state would lead to unnecessary, but also unavoidable cross-ABC coupling.


### Decorator based registration
This approach could work one of two ways.

1. acts as a class decorator, similar to `dataclasses.dataclass`, and does some magic as to mutate the class being registered.
 - this would be similar to the ABC approach, and suffer from similar issues
2.  Acts as a pass-through class decorator, similar to `@command`
 - Relatively simple to implement 
 - functions are hashable, so we can use a set for storage
 - instances  of objects cannot be stored this way
 - no ABC enforcement, would need to depend on introspection for sanity
 - could be built with native async support
 - potentially reinventing the wheel.

# `Pluggy` based approach
This approach is somewhat similar to the decorator based approach insofar as it also uses decorators, however *how* it uses them is for a completely different approach.


pluggy's interface is rather streight forward, and is made up of three parts.
1. the **plugin manager** 
2. the **specification**, 
3.  the **implementations** / **plugins**


## The Plugin manager
the plugin manager serves as a centralized instance of an object used to manage a plugin-enabled service.
Specifications and implementations are registered against this management object.

This Configuration implementation holds such an object as a static, module-bound object in `/src/config/__init__.py`

## Specifications
Specification functions are used to designate /what/ a plugin event looks like, this can be considered a prototype similar to those in c++ and similar languages.

a sample hook specification pulled from this feature `/src/config/_spec.py`
```py
rehash_spec = pluggy.HookspecMarker(_PLUGIN_NAME)

@rehash_spec
def rehash_handler(data: typing.Dict):
    """
    Apply new configuration data

    Args:
        data (typing.Dict): registered subsection of new configuration data to apply.

    """
    ...
```
 
Once hooks are defined, they can be registered against the plugin manager. 

This implementation does this directly after initializing the plugin manager in `/src/config/__init__.py`

## The Implementations AKA the good stuff AKA Usage
Plugins can take one of two forms, one can either bind a *namespace* such as a Module or an instance of an object. 

**Note** plugins need not be objects, namespaces work too! this is an improvement over the ABC approach discussed above

To register as a plugin, the object / namespace needs to be registered against the plugin manager.
this implementation exposes both the plugin manager and the necessary plugin marker for external use.

to register a plugin one simply needs to decorate the functions / methods with the exported `config_marker` decorator, then register the object itself against the manager.
```py
from src.config import config_marker, plugin_manager
class MyConfigAwareClass:
    @config_marker
    def validate_config(self, data: Dict):
        print(f"in {self}.validate_config")

    @config_marker
    def rehash_handler(self, data: Dict):
        print(f"in {self}.rehash_handler")

# make an instance 
instance = MyConfigAwareClass()

# register the plugin
plugin_manager.register(instance, "OptionalPluginName")
```

a module can be used in place of `instance`, when registering modules as plugins.
```py
from . import rat_command
...
from src.config import plugin_manager

plugin_manager.register(rat_command, "commands")
```


## limitations of this approach
1. Currently pluggy only supports synchronous plugins, and does not support an async mode of operation.


2. While pluggy works with both, there is a word of caution to be had, Pluggy's plugin-detection mechanism tends to call `getattr` on every attribute of the object, meaning something like the `Mechaclient`'s lazy loaded attributes don't play well with it.

3. we canot have multiple plugins with the same name (or hash)

# DL; DR (but please do) 
This PR adds the capability to validate and reload the configuration file without bouncing the Mechaclient (for the most part).

## Exposed events
### Validator
This event may be used by plugins to receive the candidate configuration object for validation. 
Each plugin gets the same, full, configuration dictionary.

Plugins are responsible for validating any portion of the configuration object they directly depend on, though they are allowed to validate other portions as they see fit. 

Plugins may raise either a `ValueError` or a `KeyError`, though any exception will do. (but please keep to these two, so i don't need to introduce a second wildcard exception handler 😷 )

```py
@validator_spec
def validate_config(data: typing.Dict):
    """
    Validate new configuration data.

    Args:
        data (typing.Dict): new configuration data  to validate

    Raises:
        ValueError: config section failed to validate.
    """
    ...
```

### Applicator
This event may be used by plugins to receive the new configuration object for application internally.

Each plugin gets the same, full, configuration dictionary.

Each plugin is responsible for applying any changes to its underlying data structures.

```py
@rehash_spec
def rehash_handler(data: typing.Dict):
    """
    Apply new configuration data

    Args:
        data (typing.Dict): new configuration data to apply.

    """
    ...
```

## Flow of control
the project will **no longer** load the configuration file upon import, instead differing this to the program's entry point code.

This means the configuration file is **unavailable without using the plugin hooks**.
Further, this means the config file is **unavailable until post-init** so should not be depended upon as a module / plugin loads.

theunkn0wn1 committed:
SPARK-130 Implement Base system for rehashing

theunkn0wn1 committed:
SPARK-130 Added base config handling boilerplate to the mechaclient

theunkn0wn1 committed:
Remove unintentionally commited file

theunkn0wn1 committed:
SPARK-131 Added a rehash command in the debug package

theunkn0wn1 committed:
SPARK-130 renamed implementation_marker to be more descriptive

theunkn0wn1 committed:
SPARK-130 Removed pluggy registration from mechaclient

theunkn0wn1 committed:
SPARK-130 begun adding Commands plugin stuff

theunkn0wn1 committed:
SPARK-130 WIP validator to rat_command

theunkn0wn1 committed:
SPARK-130 refactored Permission to allow null construction

theunkn0wn1 committed:
SPARK-130 Permission is now not hashable.
- Differ permission initialization to post-init

theunkn0wn1 committed:
SPARK-130 fixed remaining incompatibilities with the new config system

theunkn0wn1 committed:
Formally removed `config.config`

theunkn0wn1 committed:
SPARK-130: Adjusted DBM to fit rehash application

theunkn0wn1 committed:
Enable Galaxy to receive rehash events

shatteredbeam committed:
SPARK-130 Implement preliminary validation for DBM

theunkn0wn1 committed:
SPARK-130 implemented an actual hash check

theunkn0wn1 committed:
SPARK-130: removed debug GC call

theunkn0wn1 committed:
SPARK-130 Fix ratmama parser's config usage

theunkn0wn1 committed:
SPARK-130 give ratmama's parser a name

theunkn0wn1 committed:
SPARK-130 \n at EOL and other formatting things.

Eearslya commented on Dockerfile:
```diff
@@ -13,7 +13,7 @@
 
 # Use an official Python runtime as a parent image
-FROM python:3.6.6-alpine
+FROM python:3.7-alpine
```
[Scope] Isn't the whole Docker config on the chopping block?

theunkn0wn1 replied:
that shouldn't have made it in, il need to hunt down whatever commit touched that.

Eearslya replied:
cff1264

`@theunkn0wn1 SPARK-130 begun adding Commands plugin stuff`

theunkn0wn1 replied:
resolved via an interactive rebase. thanks @Eearslya

Eearslya commented on config.py:
```diff
@@ -99,30 +101,55 @@ def setup_logging(logfile: str):
+    """
+    Validate & apply configuration from disk
     Args:
         filename (str): path and filename to load.
```
[Style] No return documentation.

Eearslya commented on config.py:
```diff
@@ -99,30 +101,55 @@ def setup_logging(logfile: str):
+    global _hash
+
+    if file_hash == _hash:
+        raise FileExistsError(f"new file's hash matches what we already have!"
```
[Question] Why is this a fatal condition?
If we're attempting to reload an unchanged config file, shouldn't it simply return without doing anything?

theunkn0wn1 replied:
This is the only effective way to assert that we actually loaded a *new* configuration from disk, and it is a error state to reload the same configuration state. 

My reasoning here is the code handling the rehash itself would be expecting this exception to be raised and catch it, returning a well formatted error to the end users.

Eearslya replied:
I still find it a bit odd to use an exception for this use case. In my mind, if a function wanted to check if a value had changed, it would save a copy before and verify afterwards. Exceptions usually mean something has gone wrong. Arguably, reloading the same config should not be an error state *for this specific method*.

theunkn0wn1 replied:
given the hastles i had when initially testing this implementation, it *was* arguably an error in this method to load the same file twice.

theunkn0wn1 replied:
Seems neither you nor @shatteredbeam like the exception, so il remove it.
Rehashing will now be state-agnostic.

Eearslya commented on config.py:
```diff
@@ -99,30 +101,55 @@ def setup_logging(logfile: str):
+
+    if file_hash == _hash:
+        raise FileExistsError(f"new file's hash matches what we already have!"
+                              f"\told = {_hash}\tnew={file_hash}")
```
[Error][Not Blocking] If the condition is that these two are equal, isn't it a bit redundant to log both of them?

theunkn0wn1 replied:
I found it very useful to have both, even if logically speaking logging both is redundant.

Eearslya commented on config.py:
```diff
@@ -99,30 +101,55 @@ def setup_logging(logfile: str):
+    return config_dict
+
+
+def load_config(filename) -> Tuple[Dict, str]:
```
[Style] Missing docstring.

Eearslya commented on config.py:
```diff
@@ -99,30 +101,55 @@ def setup_logging(logfile: str):
+    # update hasher with the raw bytes of the file
+    hasher.update(path.read_bytes())
+
+    del infile
```
[Question] Is this not handled automatically by the `with` block?

theunkn0wn1 replied:
refering to the `del`? no.

the `with` block simply ensures the file is closed, but it does not remove the name nor the file handle.

Eearslya commented on src/__main__.py:
```diff
@@ -15,9 +15,11 @@
-from config import config
+from config import setup
 # noinspection PyUnresolvedReferences
+from src import commands
```
[Style] `[W0611(unused-import), ] Unused commands imported from src`

Eearslya commented on src/commands/debug.py:
```diff
@@ -55,3 +58,33 @@
+@command("getPlugins")
+@require_channel
+@require_permission(TECHRAT)
+async def cmd_get_plugins(context: Context):
```
[Style] Missing docstring.

theunkn0wn1 replied:
in debug module, OOS to review :P

Eearslya replied:
To be fair, so is the rehash command, and that's definitely in scope. :smile:

Eearslya commented on src/commands/debug.py:
```diff
@@ -55,3 +58,33 @@
+        await context.reply(f"unable to rehash configuration file.")
+        raise ValueError("failed to rehash configuration") from exc
+    except FileExistsError:
+        await context.reply(f"unable to rehash configuration, you sure it changed?")
```
[Suggestion] Following my comment above, I'm not sure this should be an error case.

Eearslya commented on src/config/__init__.py:
```diff
@@ -0,0 +1,36 @@
+"""
+__init__.py -   # FIXME teplates
```
[Style] FIXME
(I know you probably already know this already, but I'm just documenting it so it isn't forgotten.)

theunkn0wn1 replied:
It hasn't been forgotten, its one of the reasons this PR is in a draft state 😉

Eearslya commented on src/config/__init__.py:
```diff
@@ -0,0 +1,36 @@
+from . import _spec
+
+# create the plugin manager
+plugin_manager = pluggy.PluginManager(_PLUGIN_NAME)
```
[Style] `[C0103(invalid-name), ] Constant name "plugin_manager" doesn't conform to UPPER_CASE naming style`
(Either rename or add an ignore?)

theunkn0wn1 replied:
hmm.... this one may require discussion as it is not constant.

yes, the name is constant insofar as it always points to the same object, but the object to which it points to is not constant (due to registration activities)

Eearslya replied:
I feel like as long as the value of this particular variable (i.e. the *pointer* to what object this variable references) doesn't change, that it's still technically a constant.
It doesn't necessarily mean that absolutely nothing about it can ever change.

If we were to destroy and re-create it, changing the object ID, *then* it would be a problem.

theunkn0wn1 replied:
renamed.

Eearslya commented on src/config/__init__.py:
```diff
@@ -0,0 +1,36 @@
+
+# create implementation markers
+
+config_marker = pluggy.HookimplMarker(_PLUGIN_NAME)
```
[Style] `[C0103(invalid-name), ] Constant name "config_marker" doesn't conform to UPPER_CASE naming style`
(Either rename or add an ignore?)

theunkn0wn1 replied:
yeah, this should be named constant. will fix.

Eearslya commented on src/config/_constants.py:
```diff
@@ -0,0 +1,13 @@
+"""
+_constants -   # FIXME teplates
```
[Style] FIXME

Eearslya commented on src/config/_spec.py:
```diff
@@ -0,0 +1,46 @@
+"""
+_validator -   # FIXME teplates
```
[Style] FIXME

Eearslya commented on src/config/_constants.py:
```diff
@@ -0,0 +1,13 @@
+
+See LICENSE.md
+"""
+_PLUGIN_NAME = "mecha_config"
```
[Question] Is there a reason this is labeled a private variable?
It seems to be the exact kind of thing this module *wants* to export.

theunkn0wn1 replied:
no, external plugins shouldn't need to access the plugins name, further it is intended to only be used internally by the package's modules.

Eearslya replied:
Still feels odd to have an entire file dedicated to one constant value, and importing private vars is always a bit of a code smell imo.

theunkn0wn1 replied:
this name is required internally by other modules, to prevent duplicity i needed to place it somewhere the other internals could access. The alternative here is to duplicate the hardcoded string across multiple modules.

I don't see a reason in which something external to this plugin system needs to access this name.

shatteredbeam replied:
I agree with Eearsy here, but I can't offer a better alternative.

theunkn0wn1 replied:
If you guys can come up with a better alternative, im all ears.

Eearslya replied:
Since this is indeed used in multiple modules, I can't come up with anything much better. I had thought it was used once and only once.

shatteredbeam replied:
That being said, I'm not going to consider it blocking.  Perhaps we can improve this at a later date, but for now I don't see a reason to halt the PR (over this).

Eearslya commented on src/config/_spec.py:
```diff
@@ -0,0 +1,46 @@
+
+from ._constants import _PLUGIN_NAME
+
+validator_spec = pluggy.HookspecMarker(_PLUGIN_NAME)
```
[Style] `[C0103(invalid-name), ] Constant name "validator_spec" doesn't conform to UPPER_CASE naming style`

theunkn0wn1 replied:
*grumbles* will fix

Eearslya commented on src/config/_spec.py:
```diff
@@ -0,0 +1,46 @@
+from ._constants import _PLUGIN_NAME
+
+validator_spec = pluggy.HookspecMarker(_PLUGIN_NAME)
+rehash_spec = pluggy.HookspecMarker(_PLUGIN_NAME)
```
[Style] `[C0103(invalid-name), ] Constant name "rehash_spec" doesn't conform to UPPER_CASE naming style`

Eearslya commented on src/mechaclient.py:
```diff
@@ -12,11 +12,12 @@
 
 """
 from logging import getLogger
+from typing import Dict, ClassVar
```
[Style] `[W0611(unused-import), ] Unused Dict imported from typing`
`[W0611(unused-import), ] Unused ClassVar imported from typing`

Eearslya commented on src/mechaclient.py:
```diff
@@ -12,11 +12,12 @@
 from pydle import Client
 
-from config import config
+from src.config import config_marker
```
[Style] `[W0611(unused-import), ] Unused config_marker imported from src.config`

Eearslya commented on src/packages/commands/rat_command.py:
```diff
@@ -43,7 +43,31 @@ class NameCollisionException(CommandException):
+
+
+@config_marker
+def rehash_handler(data: Dict):
```
[Style] Missing docstring.

Eearslya commented on src/packages/commands/rat_command.py:
```diff
@@ -43,7 +43,31 @@ class NameCollisionException(CommandException):
+
+
+@config_marker
+def validate_config(data: Dict):
```
[Style] Missing docstring.

Eearslya commented on src/packages/commands/rat_command.py:
```diff
@@ -43,7 +43,31 @@ class NameCollisionException(CommandException):
+    global PREFIX
+
+    PREFIX = data['commands']['prefix']
+    LOG.debug(f"in rehash, applying new configuration data...")
```
[Error] Debug log seems a little unhelpful without any context as to WHICH rehash it's in.

theunkn0wn1 replied:
context is added by the logger itself. reviewing the log file its quite clear which module emitted this message.

Eearslya commented on src/packages/context/context.py:
```diff
@@ -10,22 +10,38 @@
+from src.config import config_marker
 from ..user import User
 
+import logging
```
[Style] `[C0411(wrong-import-order), ] standard import "import logging" should be placed before "from src.config import config_marker"`

Eearslya commented on src/packages/context/context.py:
```diff
@@ -10,22 +10,38 @@
 See LICENSE.md
 """
-from typing import Optional, Tuple, List, TYPE_CHECKING
+import typing
```
[Question] Why is this imported wholesale instead of simply adding `ClassVar` to the imports below?

theunkn0wn1 replied:
In this instance, we ended up importing an excessive number of names from this module.

Stylistically it is better after about 5 names to just import the module wholesale.

Further, I argue its more readable to have `typing.ClassVar` than `ClassVar` etc.

theunkn0wn1 replied:
I will fix the other type hints in this module to conform to the change as well, while im looking at the file.

Eearslya commented on src/packages/context/context.py:
```diff
@@ -10,22 +10,38 @@
-PREFIX = config['commands']['prefix']
+
+@config_marker
+def validate_config(data: Dict):
```
[Style] Missing docstring.

Eearslya commented on src/packages/context/context.py:
```diff
@@ -10,22 +10,38 @@
+
+
+@config_marker
+def rehash_handler(data: Dict):
```
[Style] Missing docstring.

Eearslya commented on src/packages/database/database_manager.py:
```diff
@@ -52,6 +52,53 @@ class property, and the DatabaseManage.query() method used to perform a query.
+
+    @classmethod
+    @config_marker
+    def validate_config(cls, data: Dict):
```
[Style] Missing docstring.

Eearslya commented on src/packages/database/database_manager.py:
```diff
@@ -135,3 +183,6 @@ def __init__(self,
         return list(result)
+
+
+plugin_manager.register(DatabaseManager, "Database")
```
[Question] For consistency's sake, most modules seem to be registering through their `__init__.py` files. Should we stick to that standard?

theunkn0wn1 replied:
yes. will move.

Eearslya commented on src/packages/galaxy/galaxy.py:
```diff
@@ -319,3 +335,6 @@ def __init__(self, url: str = None):
                 await self._retry_delay(retry)
+
+
+plugin_manager.register(Galaxy, "galaxy")
```
[Question] Again, the question of consistency and where modules should register.

theunkn0wn1 replied:
Again, fair point will move.

Eearslya commented on src/packages/permissions/permissions.py:
```diff
@@ -16,18 +16,34 @@
 
 
+@config_marker
+def validate_config(data: Dict):
```
[Style] Missing docstring.

Eearslya commented on src/packages/permissions/permissions.py:
```diff
@@ -16,18 +16,34 @@
+
+
+@config_marker
+def rehash_handler(data: Dict):
```
[Style] Missing docstring.

Eearslya commented on src/packages/permissions/permissions.py:
```diff
@@ -16,18 +16,34 @@
 
+@config_marker
+def validate_config(data: Dict):
+    ...
```
[Question] Is this a to-do, or is there really no validation to be done?

theunkn0wn1 replied:
TODO, missing relevant `#TODO` comment

Eearslya commented on src/packages/ratmama/ratmama_parser.py:
```diff
@@ -13,16 +13,32 @@
-from typing import Optional
+from typing import Optional, Dict
 
 import config
```
[Style] `[W0611(unused-import), ] Unused import config`

Eearslya commented on tests/conftest.py:
```diff
@@ -319,3 +320,32 @@ def test_fact_fx() -> Fact:
+
+
+@pytest.fixture(scope="session", autouse=True)
+def global_init_fx():
```
[Style] Missing docstring.

Eearslya commented on tests/unit/test_permissions.py:
```diff
@@ -19,16 +19,17 @@
 
 import src.packages.commands.rat_command as Commands
+from config import setup
+from src.packages.cli_manager import cli_manager
```
[Style] Both of these appear to be unused imports?

theunkn0wn1 replied:
quite possibly. this Pr is still a draft so imports have yet to be optimized.

Eearslya reviewed (commented):
I like the implementation, mostly a lot of style violations and needed documentation.

Of course there's plenty of testing to add as well, but you've already acknowledged that.
I'd also like to generally take a look at the grammar being used within documentation and logging. Should we care about proper capitalization and structuring sentences?

theunkn0wn1 committed:
SPARK-130 moved manager to own file, moved config from project root

theunkn0wn1 committed:
SPARK-130 plugin_manager -> PLUGIN_MANAGER

theunkn0wn1 committed:
SPARK-130 fixed a bunch of pylint errors in spec

theunkn0wn1 committed:
SPARK-130 resolved a BUNCH of pylint warnings
config_marker -> CONFIG_MARKER

theunkn0wn1 committed:
SPARK-130 more pylint stuff. cli_manager.args -> GET_ARGS

theunkn0wn1 committed:
SPARK-130 more pylint stuff.

Eearslya commented on src/__main__.py:
```diff
@@ -15,9 +15,11 @@
 
-from config import config
 # noinspection PyUnresolvedReferences
+from src import commands  # pylint: disable=unused-import
```
[Question] If this import truly is unused, why suppress the linting error?

theunkn0wn1 replied:
its not unused. 

What happens here is `src.commands` package simply *needs to be loaded*, but its contents are not *directly* used.

This package contains our user-facing Interface, the commands system. Commands are registered at load time, but that **requires them to be loaded**

This seems like the best place to do this, is at the top of main as part of pre-initialization. 

The import needs to exist, even if its not used by this module directly.

Eearslya commented on src/config/_manager.py:
```diff
@@ -0,0 +1,32 @@
+"""
+$file.fileName - $shortDescripion  # FIXME template
```
[Style] FIXME

theunkn0wn1 replied:
I argue that a `fixme` is error. will fix

Eearslya commented on src/config/_manager.py:
```diff
@@ -0,0 +1,32 @@
+#  Licensed under the BSD 3-Clause License.
+#
+#  See LICENSE.md
+#
```
[Error] Two styles of file docstring.

Eearslya commented on src/config/_parser.py:
```diff
@@ -99,30 +107,76 @@ def setup_logging(logfile: str):
+    _hash = file_hash
+
+    setup_logging(config_dict['logging']['log_file'])
+    logging.info(f"new config hash is {file_hash}")
```
[Suggestion] Demote to debug level logging?

theunkn0wn1 replied:
Im thinking it would be best to emit the fact a rehash occured, as well as the destination hash, as part of normal operation.

Perhaps the more verbose elements (such as the individual components handling things) would be debug level details.

Eearslya replied:
I agree that rehashes should be logged, but I just think that outputting the actual SHA256 is a bit overkill for normal operation.

theunkn0wn1 replied:
No kill quite like the overkill.

I would rather know the exact SHA256 of what it applied without bouncing mecha into verbose mode to find out.

Eearslya commented on src/packages/cli_manager/cli_manager.py:
```diff
@@ -2,7 +2,7 @@
 
 This module will automatically parse arguments provided from the command line when this module is
-**first imported** and provide it for importing by other modules via the `args` attribute
+**first imported** and provide it for importing by other modules via the `GET_ARGUMENTS` attribute
```
[Error] `GET_ARGUMENTS` is not an attribute, it is a callable.

theunkn0wn1 replied:
not strictly true. it is an attribute as it is a member of a (module) object.

That said, the more apt description of this object is now `method` and will be adjusted

theunkn0wn1 replied:
Resolved in last push.

Eearslya commented on src/packages/context/context.py:
```diff
@@ -10,22 +10,38 @@
+
+@CONFIG_MARKER
+def validate_config(data: Dict):
+    if not isinstance(data['commands']['prefix'], str):
```
[Question] This value is validated twice, in Context and Rat_Command. Is that intentional?

theunkn0wn1 replied:
There is no restriction on what plugin may validate what portion of a configuration file.

Eearslya replied:
I'm not saying there should be, I just find it odd that it's validated **twice**. Seems like a good place for a "changed-in-one-place-but-not-the-other" error to appear.

theunkn0wn1 replied:
Ahh, I misunderstood.

Fair point, il look at it.

theunkn0wn1 replied:
validations in rat command have been removed. issue resolved

Eearslya commented on src/packages/commands/rat_command.py:
```diff
@@ -43,7 +43,37 @@ class NameCollisionException(CommandException):
 
 # character/s that must prefix a message for it to be parsed as a command.
-PREFIX = config['commands']['prefix']
+PREFIX: str
```
[Question] If prefix handling is done in Context, why does RatCommand need its own?

theunkn0wn1 replied:
curiously enough, its used (outside configuration validation) in exactly one place within the module, and it doesn't even need to be there.

```py
                LOG.debug(f"Could not find command or rule for {PREFIX}{ctx.words[0]}.")
```

Il consider removing it.

Eearslya commented on src/packages/database/database_manager.py:
```diff
@@ -52,6 +52,53 @@ class property, and the DatabaseManage.query() method used to perform a query.
+        """
+        cls._config = data
+
+    _config: ClassVar[Dict] = {}
```
[Style] [Non-Blocking] Wouldn't this be better placed at the top of the class?

theunkn0wn1 replied:
possibly, though it makes no difference functionally. 
Il see to moving it with other pending changes.

theunkn0wn1 committed:
SPARK-130 Documentation update

theunkn0wn1 committed:
SPARK-130 more documentation
 - configuration_fx now implicitly uses the global init
  - as it is auto-use

theunkn0wn1 committed:
SPARK-130 corrected bad documentation in _manager.py

theunkn0wn1 committed:
SPARK-130 Apt descriptions are apt.

theunkn0wn1 committed:
SPARK-130 We only need to emit one hash since both are the same

theunkn0wn1 committed:
SPARK-130 removed rat_command.PREFIX

theunkn0wn1 committed:
SPARK-130 started adding more verbose documentation

theunkn0wn1 committed:
SPARK-130 Fix line length issue
 - new IDE install, who dis?

theunkn0wn1 committed:
SPARK-130 Resolved several PR review comments
 - regulated Context's type hints

theunkn0wn1 committed:
SPARK-130 Fixed access to private member in src.packages.context

theunkn0wn1 committed:
SPARK-130 Moved Galaxy's plugin registration to conform to standard

theunkn0wn1 committed:
SPARK-130 trailing whitepsace has been left behind

theunkn0wn1 committed:
SPARK-130 Resolved some lingering documentation issues

theunkn0wn1 commented:
Feature complete.

Just missing some tests and whatever you guys come up with in review

theunkn0wn1 committed:
SPARK-130 Commands doesn't currently need to be a plugin.

shatteredbeam commented on src/commands/debug.py:
```diff
@@ -55,3 +57,34 @@
+    except (KeyError, ValueError) as exc:
+        await context.reply(f"unable to rehash configuration file.")
+        raise ValueError("failed to rehash configuration") from exc
+    except FileExistsError:
```
[Suggestion]  Let's not raise during a rehash if there are no value changes.

theunkn0wn1 replied:
[error] this no longer raises, this except block should now be removed

theunkn0wn1 committed:
SPARK-130 Removed exception when reloading same configuration file
 - rehashing is now state-agnostic

shatteredbeam commented on src/mechaclient.py:
```diff
@@ -49,16 +49,24 @@ def __init__(self, *args, **kwargs):
+        self._config = mecha_config if mecha_config else {}
         super().__init__(*args, **kwargs)
 
+    @property
```
[Question]  Do we need a public property here?  I would think config should remain private across the module, since each module has its own config object now.

theunkn0wn1 replied:
No we do not, each module gets its own private property in this regard, and shouldn't be getting it from the mechaclient. will fix.

theunkn0wn1 committed:
SPARK-130 Do file loading in one go
 - eliminates potential data race
 - probably slightly speeds things up too.

shatteredbeam commented on src/packages/galaxy/galaxy.py:
```diff
@@ -68,7 +84,7 @@ class Galaxy:
 
     def __init__(self, url: str = None):
-        self.url = url or config['api']['url']
+        self.url = url or self.configuration['api']['url']
```
[Style] property name across modules should be consistent.

shatteredbeam reviewed (changes requested):
Left in RC status due to missing tests.  Yes, I know they're forthcoming.

theunkn0wn1 committed:
SPARK-130 Remove config property from the MechaClient

theunkn0wn1 committed:
SPARK-130 consistency in variable name

theunkn0wn1 committed:
SPARK-130 Fixed broken attribute reference

theunkn0wn1 committed:
SPARK-130 Fixed non-rendering codeblock in usage docs
 - removed redundant elipsis from spec

theunkn0wn1 committed:
SPARK-130 Fixed broken attribute reference

Eearslya commented on src/config/__init__.py:
```diff
@@ -0,0 +1,85 @@
+    See Also:
+        `src.config <src >`_
+
+hello world!
```
[Error] A bit of test code left in? :smile:

theunkn0wn1 replied:
yep, thats a bit of test code. Sphinx & rst can be a bit finicky on occasions.

will fix!

theunkn0wn1 commented on src/__main__.py:
```diff
@@ -41,6 +43,8 @@
     Initializes and connects the client, then passes it to rat_command.
     """
+
+    config = setup(cli_manager.GET_ARGUMENTS().config_file)
```
[question]  should i move `setup()` to something a tad more descriptive, like `reload_configuration` ?

theunkn0wn1 commented on src/mechaclient.py:
```diff
@@ -58,7 +58,7 @@ def __init__(self, *args, **kwargs):
         LOG.debug(f"Connecting to channels...")
         # join a channel
-        for channel in config["irc"]["channels"]:
+        for channel in self.config["irc"]["channels"]:
```
[error] this should now point to `self._config` as the public `.config` property was removed.

theunkn0wn1 commented on src/packages/galaxy/galaxy.py:
```diff
@@ -11,21 +11,37 @@
-from typing import Dict, List, Optional, Tuple, Union
+import logging
+from html import escape
+from typing import Dict, List, Optional, Tuple, Union, ClassVar
```
[style] this should be moved to a wholesale typing import due to number of imports.

theunkn0wn1 commented on src/packages/galaxy/galaxy.py:
```diff
@@ -11,21 +11,37 @@
+            data (Dict): new configuration data to apply.
+
+        """
+        cls.config = data
```
[error][style] this should be a private member

theunkn0wn1 reviewed (commented):
Did a cursory review over my own work looking for anything obvious that can be improved, found a few minor things. 

changes pending!

Eearslya commented on src/packages/database/database_manager.py:
```diff
@@ -53,6 +53,59 @@ class property, and the DatabaseManage.query() method used to perform a query.
+
+        # Host
+        if not isinstance(module_config['host'], str):
+            raise ValueError("host must be a string.")
```
[Question] Perhaps these error messages should be a tad more specific? Unlike logging, we don't have any context here.
We could possibly have several config options by the name of `host` under different categories.

theunkn0wn1 replied:
the erroring module will be in the stacktrace, though the most direct thing we can do here is have `__name__` embedded into the error strings or something.

Thoughts?

Eearslya commented on src/packages/database/database_manager.py:
```diff
@@ -53,6 +53,59 @@ class property, and the DatabaseManage.query() method used to perform a query.
+
+        # Port
+        if not isinstance(module_config['port'], int):
+            raise ValueError("port must be an integer.")
```
[Question] Might it be better to keep a running list of errors and return them at the end? The fast-fail solution we have now means that users are only made aware of one config issue at a time.

theunkn0wn1 replied:
As per [pluggy's documentation](https://pluggy.readthedocs.io/en/latest/#exception-handling), 
> If any hookimpl errors with an exception no further callbacks are invoked and the exception is packaged up and delivered to any Wrappers before being re-raised at the hook invocation point:

So holding for several errors in validation is not possible.

Or are you referring to within each module have it gather exceptions and raise one at the end?

Eearslya commented on src/packages/permissions/permissions.py:
```diff
@@ -16,18 +16,51 @@
+    __slots__ = ["level", "_vhosts", "_denied_message"]
 
-    def __init__(self, level: int, vhosts: Set[str],
+    def __init__(self, level: int = -1, vhosts: Set[str] = None,
```
[Style] I've always had it drilled into my head that magic numbers are bad. If `-1` is meant to mean something special, it should be assigned to a constant variable that can then be documented.

theunkn0wn1 replied:
Its not really a magic number, and i feel it requires no documentation.

Without specific initialization, all permission objects have a default permission level of -1, which is lower than any assigned value. 

I suppose this is a "fail safe" default, rather than a "fail secure".
Optimally, this value is immediately overwritten during post-init & during subsequent configuration file reloads.

My question here is should we "fail secure" and set this to be a very large integer value rather than a out-of-range small integer.

Eearslya commented on tests/unit/test_configuration.py:
```diff
@@ -0,0 +1,10 @@
+
+def test_rehash(configuration_fx):
+    PLUGIN_MANAGER.hook.rehash_handler(data=configuration_fx)
+def test_validate(configuration_fx):
```
[Style] Newline between method definitions.

theunkn0wn1 replied:
this file wasn't supposed to be commited yet... sigh.

theunkn0wn1 committed:
SPARK-130 remove outdated exception check, refine wording of rehash
 - also adjusted `setup` to return the data and the hash

theunkn0wn1 committed:
SPARK-130 modified galaxy's type hints to match style

theunkn0wn1 committed:
SPARK-130 fixed bad reference to config in mechaclient

theunkn0wn1 committed:
SPARK-130 made galaxy's config attribute private

theunkn0wn1 committed:
SPARK-130 removed test code.

theunkn0wn1 committed:
SPARK-130 Add debug commands module to coverage blacklist
 - not production code, testing arguably isn't required here.

theunkn0wn1 committed:
SPARK-130 begun adding validation tests

theunkn0wn1 committed:
SPARK-130 Added database validation tests
 - adjusted some bad checks in DBM

theunkn0wn1 committed:
SPARK-130 added a missing check to DBM test

theunkn0wn1 committed:
SPARK-130 pack DBM tests properly...

theunkn0wn1 committed:
SPARK-130 values not items

theunkn0wn1 committed:
SPARK-130 Add permissions validation checks

theunkn0wn1 committed:
SPARK-130 synchronized DBM's typing imports
 - fixed a bad isinstance check

theunkn0wn1 committed:
Merge branch 'SPARK-130' of https://github.com/theunkn0wn1/pipsqueak3 into SPARK-130

Eearslya commented on src/commands/debug.py:
```diff
@@ -55,3 +57,33 @@
+@command("rehash")
+@require_channel(message="please do this where everyone can see 😒")
+@require_permission(TECHRAT, override_message="no.")
+async def cmd_rehash(context: Context):
```
[Question] Should this really be among the debug commands?

As I understand it, the debug commands are the ones not meant to ever be used in production, but this one certainly would be, as it's half of the point of the PR.
Am I mis-understanding the purpose of this file?

theunkn0wn1 replied:
you are misunderstanding the purpose of this file.

I  use it as a playground to experiment with commands-based functionality, not for final-product implementations.

That said, this command doesn't require much rework to be usable in production, so it proably should be moved out of this module

theunkn0wn1 replied:
moved the command to a new commands module for administration commands.
I foresee other similar commands being implemented in the future for administrative tasks so i am gunna just make a module for them now.

Eearslya reviewed (dismissed):
There are one or two aspects about this I'm not fully confident in yet, and documentation cleanup will need to be done in another pass.

But at the same time, if we sit here and polish out absolutely everything, this will never be merged. The code presented is functional and all of my concerns can easily be addressed without severe refactors later on.

For now, I'm giving the green light.

theunkn0wn1 committed:
SPARK-130 Moved rehash command to administration commands module
 - also made administration commands module a thing.

theunkn0wn1 committed:
SPARK-130 finished rehash command tests
 - removed test_configuration as its unneeded
 - individual plugins are responsible for their own tests.

theunkn0wn1 committed:
SPARK-130 Added test for "success" behavior
 - because coverage -_-

shatteredbeam commented on tests/unit/commands/test_administration.py:
```diff
@@ -0,0 +1,65 @@
+
+@pytest.mark.asyncio
+async def test_rehash_not_authorized(bot_fx, async_callable_fx, monkeypatch):
+    context = await Context.from_message(bot_fx, "#unittest", "some_recruit", "!rehash 👿")
```
Hardcoded channel name?

theunkn0wn1 replied:
This is within a test file, im not sure i see the issue with using a hardcoded channel name?

The channel is ultimately irrelevant, it simply needs to be a valid channel name.

Further, in many test files we make use of similar hardcoded strings, why are you calling these out in specific?

shatteredbeam commented on tests/unit/commands/test_administration.py:
```diff
@@ -0,0 +1,65 @@
+    """
+    verifies the commands behavior when the setup() call fails (predictably).
+    """
+    context = await Context.from_message(bot_fx, "#unittest", "some_admin", "!rehash")
```
Another hard coded channel name.

shatteredbeam reviewed (changes requested):
Other than pointing out hardcoded channels in the test file, Most of this has already been discussed.
